Source code level attribution in profiler

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43898 Source code level attribution in profiler**

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

https://gist.github.com/ilia-cher/9c7901670fb48082af92386b1889e2fe
https://gist.github.com/ilia-cher/69a01a6ef294511a797e8cf4a1e6db75

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

Commit:
[wip] Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Name                                 Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Source Location
-----------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  --------------------------------------------
ts_method_1                          10.43%           235.364us        36.46%           822.920us        822.920us        1                test/test_profiler.py(70): test_source
aten::add                            7.52%            169.833us        8.88%            200.439us        200.439us        1                test/test_profiler.py(69): test_source
aten::normal_                        6.26%            141.380us        6.26%            141.380us        141.380us        1                test/test_profiler.py(67): test_source
aten::add                            5.80%            130.830us        8.41%            189.800us        63.267us         3                test/test_profiler.py(72): test_source
aten::sum                            5.02%            113.340us        8.39%            189.475us        189.475us        1                test/test_profiler.py(64): ts_method_1
aten::add                            4.58%            103.346us        6.33%            142.847us        142.847us        1                test/test_profiler.py(62): ts_method_1
aten::mul                            4.05%            91.498us         9.62%            217.113us        217.113us        1                test/test_profiler.py(71): test_source
aten::add                            4.03%            90.880us         5.60%            126.405us        126.405us        1                test/test_profiler.py(58): ts_method_2
aten::empty                          3.49%            78.735us         3.49%            78.735us         19.684us         4                test/test_profiler.py(72): test_source
Reviewers:

ngimel commented on test/test_profiler.py:
```diff
@@ -44,5 +44,48 @@ def test_mem_leak(self):
+            return w.sum()
+
+        with profile(record_shapes=False, with_source=True) as p:
+            x = torch.randn(10, 10, requires_grad=True)
```
please add tests for nn.Module functions and composite functions such as matmul

ngimel commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -274,6 +288,66 @@ struct ProfilerThreadLocalState
+      }
+      // in some cases we won't find the corr. forward node (e.g. GraphRoot)
+    } else {
+      auto ts_loc = torchScriptCallSite();
```
please leave a comment here regarding under which conditions ts_loc and py_stack are necessary (ts_loc for scripted, py_stack for eager, right?), and try not to call py_stack in scripted mode

ngimel reviewed (commented):
Please benchmark the added overheads and make sure it's acceptable with disabled profiling/profiling w/o source attribution.

dzhulgakov commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -274,6 +288,66 @@ struct ProfilerThreadLocalState
+  std::vector<std::string> pythonCallstack() {
+    std::vector<std::string> py_stack;
+
+    PyGILState_STATE gil_state = PyGILState_Ensure();
```
have we timed how slow is it to decode stuff to strings? it might be worthwhile to just bump the refcount on the `f_code` objects and keep them around and decode in post-processing. That's what lsprof seems to do: https://github.com/python/cpython/blob/master/Modules/_lsprof.c#L108 and the processing in https://github.com/python/cpython/blob/master/Lib/cProfile.py#L122

keeping the pointer to the entire PyFrameObject is probably not cool as it might keep all locals vars around

cc @zdevito as a resident cpython expert

dzhulgakov replied:
also you could try to write it with pybind - might be a bit more readable, up to you

dzhulgakov commented on torch/csrc/autograd/profiler.h:
```diff
@@ -117,6 +120,11 @@ struct TORCH_API ProfilerConfig {
 
 };
 
+struct EventSourceLocation {
```
just curious - why one is string and another is list? for uniform post-processing it makes sense to use the same format

ilia-cher replied:
needs to be upd

dzhulgakov commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1173,17 +1232,21 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
           } break;
-          case OP:
+          case OP: {
+            TLSCallInfoGuard g(call_info);
```
I'm a bit worried about introducing another TLS write on every op. Given that we have Frame hierarchy in TS too (just like we do in python) - can't we instead just have a TLS of "active interpreter" (which gets updated only once in runImpl) and then have a method to obtain from the active frame and it's hierarchy)? I think we'd need to move `af.pc` into `frame.pc` every time (why are they different in the first place?), but that should be much smaller overhead than hitting a threadlocal

ilia-cher replied:
active interpreter TLS pointer - like this idea too, i considered doing this though that would expose some parts of the JIT interpreter, but it could also simplify the rest of the code

dzhulgakov commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1044,28 +1067,62 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+      // if that's the first time we activate the frame, start
+      // the record function
+      if (frame.pc == 0) {
+        checkAndStartRecordFunction(frame, stack);
```
it's weird that you put it into ActiveFrame constructor. Seems like `runGraphFunction` is a better place where we actually enter a new function. Or is it to make it work with "suspendible" tasks? Then I'm afraid it won't generate the nested events beyond the first level nesting anyway

ilia-cher replied:
i moved it out of runGraphFunction because there are other cases when we'd want to call RecordFunction, e.g. when we do fork (vs in-script function/method call)

passing stack into ActiveFrame - that's purely for RecordFunction in cases when this is the first time we execute the frame,
and it guarantees we don't forget to do it,
originally i just called checkAndStartRecordFunction after the several use cases of ActiveFrame, we could return to that

dzhulgakov reviewed (commented):
@suo @zdevito on TorchScript changes. I left some comments, but better suggestions are welcomed

Perf measurements are absolutely a must. Both when source attribution is off and when it's on

albanD commented on torch/csrc/autograd/function.h:
```diff
@@ -104,6 +104,7 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
       uint64_t sequence_nr,
       edge_list&& next_edges = edge_list())
       : sequence_nr_(sequence_nr),
+      thread_id_(at::RecordFunction::currentThreadId()),
```
It's not something I like to see for sure. Though I never actually did a full benchmark of the impact on memory and speed.

Do you actually always need this thread id or just when the profiler is running?
If so, I think a nice way to guard this to only run when the profiler runs and store it in the metadata struct that we also use for anomaly mode to store this (you can make the python dict initialization lazy to improve perf when you just store and read the thread_id).


> I'm not sure how in general sequence_nrs work - are they thread local?

Yes these are just thread local variables (not propagated by TLSState)

ilia-cher replied:
sequence numbers are thread local, that's why i needed to add original fwd thread id, otherwise the fwd-backward matching can be incorrect

ilia-cher replied:
we then should either set or not set both of them, because seq. number without thread is not very useful;
if extra int64 is too much I'd better go with packing though - less runtime checks

albanD replied:
The autograd engine make heavy use of the seq number (that does not require the thread). So we have to set this one.

Commit:
Update on "[wip] Source code level attribution in profiler"

[wip] Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Name                                 Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Source Location
-----------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  --------------------------------------------
ts_method_1                          10.43%           235.364us        36.46%           822.920us        822.920us        1                test/test_profiler.py(70): test_source
aten::add                            7.52%            169.833us        8.88%            200.439us        200.439us        1                test/test_profiler.py(69): test_source
aten::normal_                        6.26%            141.380us        6.26%            141.380us        141.380us        1                test/test_profiler.py(67): test_source
aten::add                            5.80%            130.830us        8.41%            189.800us        63.267us         3                test/test_profiler.py(72): test_source
aten::sum                            5.02%            113.340us        8.39%            189.475us        189.475us        1                test/test_profiler.py(64): ts_method_1
aten::add                            4.58%            103.346us        6.33%            142.847us        142.847us        1                test/test_profiler.py(62): ts_method_1
aten::mul                            4.05%            91.498us         9.62%            217.113us        217.113us        1                test/test_profiler.py(71): test_source
aten::add                            4.03%            90.880us         5.60%            126.405us        126.405us        1                test/test_profiler.py(58): ts_method_2
aten::empty                          3.49%            78.735us         3.49%            78.735us         19.684us         4                test/test_profiler.py(72): test_source
Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "[wip] Source code level attribution in profiler"

[wip] Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Name                                 Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Source Location
-----------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  --------------------------------------------
ts_method_1                          10.43%           235.364us        36.46%           822.920us        822.920us        1                test/test_profiler.py(70): test_source
aten::add                            7.52%            169.833us        8.88%            200.439us        200.439us        1                test/test_profiler.py(69): test_source
aten::normal_                        6.26%            141.380us        6.26%            141.380us        141.380us        1                test/test_profiler.py(67): test_source
aten::add                            5.80%            130.830us        8.41%            189.800us        63.267us         3                test/test_profiler.py(72): test_source
aten::sum                            5.02%            113.340us        8.39%            189.475us        189.475us        1                test/test_profiler.py(64): ts_method_1
aten::add                            4.58%            103.346us        6.33%            142.847us        142.847us        1                test/test_profiler.py(62): ts_method_1
aten::mul                            4.05%            91.498us         9.62%            217.113us        217.113us        1                test/test_profiler.py(71): test_source
aten::add                            4.03%            90.880us         5.60%            126.405us        126.405us        1                test/test_profiler.py(58): ts_method_2
aten::empty                          3.49%            78.735us         3.49%            78.735us         19.684us         4                test/test_profiler.py(72): test_source
Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "[wip] Source code level attribution in profiler"

[wip] Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "[wip] Source code level attribution in profiler"

Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "Source code level attribution in profiler"

Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "Source code level attribution in profiler"

Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

ilia-cher commented:
addressed most comments, will upd with a new round of benchmarking

dzhulgakov commented on torch/autograd/profiler.py:
```diff
@@ -1213,9 +1294,21 @@ def append(s):
 
+        if has_stack:
+            empty_headers = [""] * (len(headers) - 1)
+            for entry in evt.stack[1:MAX_STACK_ENTRY]:
```
should it be whatever value is passed to group_by_n? or is it only used when there's no grouping?

also, why does it slice the first element?

ilia-cher replied:
it slices the first element bc we already outputted it, this code outputs table row by row and we use multiple rows per stacktrace

ilia-cher replied:
evt.stack here is either the original stack of the event, or (when we use group by group_by_n) an averaged event which has it's stack already limited to the first group_by_n
this MAX_STACK_ENTRY check is a sanity check to avoid very long profiler outputs

ilia-cher replied:
evt.stack is limited to top group_by_n when we do group by (key_averages);
this MAX_STACK_ENTRY check is a sanity check to avoid very long profiler outputs

dzhulgakov commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -202,31 +203,41 @@ struct ProfilerThreadLocalState
+
+      // backward nodes source range corresponds to the forward node
+      if (config_.with_source && fn.scope() != at::RecordScope::BACKWARD_FUNCTION) {
+        auto cs = jit::currentCallstack();
```
I guess if we want to be even fancier - we could unroll the C++ callback (c10::Error has the code for that), then if we encounter any Torchscript in there - append the TS callback and if any python - python callback. That'd make this profiler work even with C++ frontend or with non-trivial custom ops. Downside - it might be spammy with various PT internals from the dispatcher.

I'm not insisting we do it right away, but maybe leave a comment.

dzhulgakov reviewed (commented):
These changes to TS look better, but @suo to double check

Commit:
Update on "Source code level attribution in profiler"

Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "Source code level attribution in profiler"

Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

Commit:
Update on "Source code level attribution in profiler"

Source code level attribution in profiler

Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output:
https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Reviewers:

gh-metadata: pytorch pytorch 43898 gh/ilia-cher/40/head

ilia-cher commented:
added profiler benchmark - as expected when using with_stack=True it is slower, and the slow down is less when using TorchScript

ilia-cher commented:
also added comparison with the trunk to the test section

ilia-cher commented:
cc. @suo @jamesr66a to take a look over the interpreter part

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1176,10 +1185,12 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
             runGraphFunction(stack, &f, &af);
           } break;
           case OP:
+            frames.back().pc = af.pc;
```
The point of the active frame is to avoid having to edit the stack on each instruction. Either this change should not be made, or the entire ActiveFrame stuff should simply be removed because it no longer serves a purpose.

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1153,6 +1158,10 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
 
     ActiveFrame af(frames.back());
+    TLSCurrentInterpreterGuard g(this);
+    if (af.pc == 0) {
```
`af.pc == 0` is a fragile check for the start of a function.  The bytecode generator might produce bytecode that jumps back to `pc == 0` so there are some rare scenarios where its 0 but the start of the function.

zdevito replied:
The code right above this does this using `stack_start_` which is more stable.

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1611,6 +1646,27 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   }
 };
 
+std::vector<FileLineFunc> currentCallstack() {
```
This couples the profilers FileLineFunc struct to the interpreter in a way that doesn't seem necessary. FileLineFunc seems like a duplicate of StackEntry here. The more flexible format for the interpreter to return is the StackEntry list. The interpreter should just return this list, and if a client like the profiler wants to translated into a simpler form like below it should do it there.

zdevito commented on torch/csrc/jit/runtime/interpreter.h:
```diff
@@ -5,6 +5,7 @@
 
 #include <ATen/ThreadLocalState.h>
 #include <ATen/core/ivalue.h>
+#include <torch/csrc/jit/frontend/source_range.h>
```
doesn't seem used

zdevito reviewed (commented):
Some comments on the interpreter changes. If we want these changes, we should remove the ActiveFrame struct and just keep a pointer to the Frame object that is active, since saving the pc on every instruction removes its purpose.

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark: https://gist.github.com/ilia-cher/1ef60eb686ec49af8e405c03e6967e7f
https://gist.github.com/ilia-cher/79870d9be7fb5454c9059b19710fd58d

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark: https://gist.github.com/ilia-cher/1ef60eb686ec49af8e405c03e6967e7f
https://gist.github.com/ilia-cher/79870d9be7fb5454c9059b19710fd58d

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -517,7 +517,7 @@ void RequestCallbackNoPython::processRpc(
                 if (cuda_profiling_enabled &&
                     0 == strcmp(e.name(), "__cuda_start_event")) {
-                  e.setCudaUs(e.cpu_us());
+                  e.setCudaUs(e.cpuUs());
```
just fyi, this might have some merge conflict with https://github.com/pytorch/pytorch/pull/44664 (which is just moving this function to a different file, so should be an easy fix)

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark: https://gist.github.com/ilia-cher/1ef60eb686ec49af8e405c03e6967e7f
https://gist.github.com/ilia-cher/79870d9be7fb5454c9059b19710fd58d

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

ilia-cher commented:
addressed @zdevito comments

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark: https://gist.github.com/ilia-cher/1ef60eb686ec49af8e405c03e6967e7f
https://gist.github.com/ilia-cher/79870d9be7fb5454c9059b19710fd58d

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

ilia-cher commented:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

ilia-cher commented:
lint,
also cc. @ngimel @dzhulgakov @robieta @suo @zdevito @jamesr66a - ready for another look

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

ngimel commented on torch/autograd/profiler.py:
```diff
@@ -208,29 +237,40 @@ def export_chrome_trace(self, path):
+        Arguments:
+            group_by_input_shapes: The key would become
+            (event name, input dimensions) rather than just event name.
+            This is useful to see which dimensionality contributes to the runtime
```
nit: which input shapes. I know it's not this PR, but still, dimensionality is the wrong work here.

ngimel commented on torch/autograd/profiler.py:
```diff
@@ -208,29 +237,40 @@ def export_chrome_trace(self, path):
+            group_by_input_shapes: The key would become
+            (event name, input dimensions) rather than just event name.
+            This is useful to see which dimensionality contributes to the runtime
+            the most and may help with dimension specific optimizations or
```
size-specific

ngimel commented on torch/autograd/profiler.py:
```diff
@@ -274,8 +314,11 @@ class profile(object):
 
         profile_memory (bool, optional): Whether to report memory usage, default: ``False``
 
+        with_stack (bool, optional): record source information for the ops
```
record source information (file and line number) for the ops

dzhulgakov commented on torch/csrc/jit/python/python_tracer.cpp:
```diff
@@ -23,25 +23,44 @@ namespace tracer {
-  size_t source_line = 0;
-  std::stringstream stack_trace;
-
+std::vector<StackEntry> pythonCallstack() {
```
nit: it's a bit weird that it lives in python_tracer (kind of unrelated), maybe move to some more generic file? (not critical)

ilia-cher replied:
i'll check if there's a better location available, if not we'd prob. better not create a new one..

dzhulgakov reviewed (commented):
Hm, your overhead measurements are actually show 6-8% slow perf with no profiling:
eager: 0.895ms -> 0.965ms
script: 0.585ms -> 0.625ms

That feels actually like a lot. Did you bisect where is it coming from? For eager mode there's even no interpreter changes - it must be from the RecordFunction itself then?

ngimel reviewed (commented):
I have small docs suggestions, other than that 6% overhead does seem like a lot, but it's not obvious where it's coming from.

ilia-cher commented:
> Hm, your overhead measurements are actually show 6-8% slow perf with no profiling:
> eager: 0.895ms -> 0.965ms
> script: 0.585ms -> 0.625ms
> 
> That feels actually like a lot. Did you bisect where is it coming from? For eager mode there's even no interpreter changes - it must be from the RecordFunction itself then?

Re. runtimes - I included them mainly to show the rough difference between with stack / without stack profiling, and python vs TS callstacks, but for other purposes it is noisy, so I included valgrind based comparison for base vs trunk default mode.

Also reran benchmark (1000 times, no profiling, PR vs base, eager/script, different machine):
```
USE_MKLDNN=1 BLAS=MKL BUILD_BINARY=1 USE_CUDA=0 python setup.py develop install --cmake

pr_eager: (each line - 1000 runs of benchmark)
Overall: avg 0.811 ms, median: 0.773, min: 0.739, stddev: 0.136 ms
Overall: avg 0.774 ms, median: 0.768, min: 0.746, stddev: 0.030 ms
Overall: avg 0.837 ms, median: 0.742, min: 0.716, stddev: 0.226 ms
Overall: avg 0.870 ms, median: 0.736, min: 0.712, stddev: 0.266 ms
Overall: avg 0.786 ms, median: 0.779, min: 0.720, stddev: 0.031 ms

pr_script:
Overall: avg 0.649 ms, median: 0.552, min: 0.527, stddev: 0.156 ms
Overall: avg 0.546 ms, median: 0.540, min: 0.510, stddev: 0.025 ms
Overall: avg 0.575 ms, median: 0.537, min: 0.520, stddev: 0.108 ms
Overall: avg 0.542 ms, median: 0.535, min: 0.521, stddev: 0.028 ms
Overall: avg 0.585 ms, median: 0.543, min: 0.521, stddev: 0.110 ms

base_eager:
Overall: avg 0.764 ms, median: 0.751, min: 0.731, stddev: 0.042 ms
Overall: avg 0.808 ms, median: 0.782, min: 0.765, stddev: 0.115 ms
Overall: avg 0.884 ms, median: 0.844, min: 0.811, stddev: 0.151 ms
Overall: avg 0.802 ms, median: 0.791, min: 0.775, stddev: 0.033 ms
Overall: avg 0.815 ms, median: 0.806, min: 0.776, stddev: 0.031 ms

base_script:
Overall: avg 0.706 ms, median: 0.588, min: 0.571, stddev: 0.184 ms
Overall: avg 0.590 ms, median: 0.582, min: 0.571, stddev: 0.031 ms
Overall: avg 0.587 ms, median: 0.580, min: 0.570, stddev: 0.030 ms
Overall: avg 0.637 ms, median: 0.596, min: 0.583, stddev: 0.109 ms
Overall: avg 0.702 ms, median: 0.612, min: 0.584, stddev: 0.162 ms

https://gist.github.com/ilia-cher/a3974d77664628f0f42d27b616738013
```

So there's no stat. sig. information about runtime for base vs PR, though callgrind stats seem to give reproducible stats (https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1) e.g. they show increase in tls accesses (expected bc of `if (profiler::profilerEnabled()) { thread_id_ = at::RecordFunction::currentThreadId(); }` and some rel. small increase in cpu instructions.

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

ilia-cher commented:
rest of comments + added support for @robieta Timer into profiler_bench.py

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

ilia-cher commented:
example of benchmarking with `--use_timer` (also varies)
```
<torch.utils._benchmark.utils.common.Measurement object at 0x7f2cc961ee80>
payload()
  Median: 787.50 us
  IQR:    62.62 us (766.31 to 828.93)
  111 measurements, 1000 runs per measurement, 1 thread
<torch.utils._benchmark.utils.common.Measurement object at 0x7fd6fde9fe80>
payload()
  Median: 913.64 us
  IQR:    33.24 us (903.73 to 936.96)
  105 measurements, 1000 runs per measurement, 1 thread
<torch.utils._benchmark.utils.common.Measurement object at 0x7f721a1e0e80>
payload()
  Median: 768.92 us
  IQR:    22.46 us (759.44 to 781.90)
  127 measurements, 1000 runs per measurement, 1 thread
<torch.utils._benchmark.utils.common.Measurement object at 0x7f6d0490de80>
payload()
  Median: 821.17 us
  IQR:    12.48 us (815.65 to 828.13)
  122 measurements, 1000 runs per measurement, 1 thread
<torch.utils._benchmark.utils.common.Measurement object at 0x7f038ce2de80>
payload()
  Median: 762.76 us
  IQR:    38.57 us (752.31 to 790.88)
  122 measurements, 1000 runs per measurement, 1 thread
```

robieta commented:
> example of benchmarking with `--use_timer` (also varies)
> 
> ```
> <torch.utils._benchmark.utils.common.Measurement object at 0x7f2cc961ee80>
> payload()
>   Median: 787.50 us
>   IQR:    62.62 us (766.31 to 828.93)
>   111 measurements, 1000 runs per measurement, 1 thread
> <torch.utils._benchmark.utils.common.Measurement object at 0x7fd6fde9fe80>
> payload()
>   Median: 913.64 us
>   IQR:    33.24 us (903.73 to 936.96)
>   105 measurements, 1000 runs per measurement, 1 thread
> <torch.utils._benchmark.utils.common.Measurement object at 0x7f721a1e0e80>
> payload()
>   Median: 768.92 us
>   IQR:    22.46 us (759.44 to 781.90)
>   127 measurements, 1000 runs per measurement, 1 thread
> <torch.utils._benchmark.utils.common.Measurement object at 0x7f6d0490de80>
> payload()
>   Median: 821.17 us
>   IQR:    12.48 us (815.65 to 828.13)
>   122 measurements, 1000 runs per measurement, 1 thread
> <torch.utils._benchmark.utils.common.Measurement object at 0x7f038ce2de80>
> payload()
>   Median: 762.76 us
>   IQR:    38.57 us (752.31 to 790.88)
>   122 measurements, 1000 runs per measurement, 1 thread
> ```

And these are 100 second runs each? If so, that's a noisy system.

ilia-cher commented:
> > example of benchmarking with `--use_timer` (also varies)
> > ```
> > <torch.utils._benchmark.utils.common.Measurement object at 0x7f2cc961ee80>
> > payload()
> >   Median: 787.50 us
> >   IQR:    62.62 us (766.31 to 828.93)
> >   111 measurements, 1000 runs per measurement, 1 thread
> > <torch.utils._benchmark.utils.common.Measurement object at 0x7fd6fde9fe80>
> > payload()
> >   Median: 913.64 us
> >   IQR:    33.24 us (903.73 to 936.96)
> >   105 measurements, 1000 runs per measurement, 1 thread
> > <torch.utils._benchmark.utils.common.Measurement object at 0x7f721a1e0e80>
> > payload()
> >   Median: 768.92 us
> >   IQR:    22.46 us (759.44 to 781.90)
> >   127 measurements, 1000 runs per measurement, 1 thread
> > <torch.utils._benchmark.utils.common.Measurement object at 0x7f6d0490de80>
> > payload()
> >   Median: 821.17 us
> >   IQR:    12.48 us (815.65 to 828.13)
> >   122 measurements, 1000 runs per measurement, 1 thread
> > <torch.utils._benchmark.utils.common.Measurement object at 0x7f038ce2de80>
> > payload()
> >   Median: 762.76 us
> >   IQR:    38.57 us (752.31 to 790.88)
> >   122 measurements, 1000 runs per measurement, 1 thread
> > ```
> 
> And these are 100 second runs each? If so, that's a noisy system.

yes, it is 100s

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

https://gist.github.com/ilia-cher/9c7901670fb48082af92386b1889e2fe

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

https://gist.github.com/ilia-cher/9c7901670fb48082af92386b1889e2fe

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

robieta reviewed (commented):
Perf looks good; variation seems to just be noise. I'll defer to others about the rest of the changes.

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

https://gist.github.com/ilia-cher/9c7901670fb48082af92386b1889e2fe

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]

Commit:
Update on "Source code level attribution in profiler"


Summary:
Adding with_source parameter to enable tracking source code
(filename and line) in profiler for eager, torchscript and autograd
modes

Test Plan:
python test/test_profiler.py

Output: https://gist.github.com/ilia-cher/bd13a3ab0c4905ce58b0079df66711fa

Benchmark:
updated benchmark: profiler https://gist.github.com/ilia-cher/ff7064a311790590ffa8445c95d4d45b
as expected, no change when no stacks are used, slower when stacks are used (in particular python is slower than TS callstack)

overhead: https://gist.github.com/ilia-cher/8a2861863d3033b56181d23ba4cfbda1
overall as expected:
small increase in cpu instructions (0.7%), small increase in tls accesses expected (bc of function.h change)

https://gist.github.com/ilia-cher/9c7901670fb48082af92386b1889e2fe

Differential Revision: [D23432664](https://our.internmc.facebook.com/intern/diff/D23432664)

[ghstack-poisoned]
